delegate-app/Main.hs

```diff
@@ -1,49 +1,133 @@
 module Main (main) where
 
 -- Prelude imports
+
+-- Prelude imports
+import Hydra.Cardano.Api.Prelude (lookupEnv)
+import Hydra.Prelude (readMaybe, traverse_)
 import Prelude
 
 -- Haskell imports
-
 import Control.Concurrent (threadDelay)
-import Control.Monad.IO.Class (MonadIO (liftIO))
+import Control.Monad (forever, (>=>))
+import Control.Monad.Extra (whenJust)
+import Control.Monad.State (MonadState (get), StateT, evalStateT, put)
+import Control.Monad.Trans
+import Control.Tracer (Tracer, contramap, stdoutTracer)
+import Data.Aeson (decode, encode)
+import Data.Maybe (fromMaybe)
+import Network.HTTP.Types (status200)
+import Network.Wai (Application, responseLBS)
+import Network.Wai.Handler.Warp (run)
+import Network.Wai.Handler.WebSockets (websocketsOr)
+import Network.WebSockets
+import Prettyprinter (Pretty (pretty))
+
+-- Hydra imports
+import Hydra.Network
 
 -- HydraAuction imports
 import HydraAuction.Delegate (
   DelegateEvent (Start),
-  DelegateInput (DelegateEvent),
+  DelegateInput (DelegateEvent, FrontendRequest),
   DelegateRunnerT,
   delegateStep,
   execDelegateRunnerT,
  )
+import HydraAuction.Delegate.Server (
+  DelegateError (FrontendNoParse),
+  DelegateServerConfig (DelegateServerConfig, dlgt'host, dlgt'port, dlgt'tick),
+  DelegateServerLog (DelegateError, DelegateOutput, FrontendConnected, FrontendInput, Started),
+  MonadTracer (trace),
+  ServerAppT,
+  TracerT,
+  askTracer,
+  dlgt'tick,
+  runWithTracer',
+ )
 
-eventsProducer :: Bool -> DelegateRunnerT IO (Maybe DelegateEvent)
+eventsProducer :: Monad m => Bool -> DelegateRunnerT m (Maybe DelegateEvent)
 eventsProducer thisIsFirstRun =
   if thisIsFirstRun
     then return $ Just Start
     else return Nothing
 
-consumer :: Maybe DelegateInput -> DelegateRunnerT IO ()
-consumer mInput = case mInput of
-  Just input -> do
-    delegateResponse <- delegateStep input
-    -- FIXME: send response to client
-    liftIO $
-      putStrLn $
-        "Delegate responses for input: " <> show delegateResponse
-  Nothing -> return ()
+delegateServerApp ::
+  (MonadIO m, MonadTracer DelegateServerLog m) =>
+  Int ->
+  ServerAppT m
+delegateServerApp tick pending = do
+  connection <- liftIO $ acceptRequest pending
+  trace FrontendConnected
+  -- FIXME: we will probably have to do this in parallel, but how can we enqueue in the same
+  -- state? probably have to remodel this
+  execDelegateRunnerT $ flip evalStateT True $ mkRunner tick connection
+
+-- FIXME: remove concrete monads and make this tagless final (needs MonadDelegate)
```

Comment:
Why remove concrete monads? This should just be called from main, no?

Reply:
I think this was mainly because it was easier to read the tagless final form than having concrete monads in this case. also we don't lose anything (perhaps performance but I don't even think that's the case here)